Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSPatcher.mm

```diff
@@ -0,0 +1,160 @@
+combining it with any of RAD Game Tools Bink SDK, Autodesk 3ds Max SDK,
+NVIDIA PhysX SDK, Microsoft DirectX SDK, OpenSSL library, Independent
+JPEG Group JPEG library, Microsoft Windows Media SDK, or Apple QuickTime SDK
+(or a modified version of those libraries),
+containing parts covered by the terms of the Bink SDK EULA, 3ds Max EULA,
+PhysX SDK EULA, DirectX SDK EULA, OpenSSL and SSLeay licenses, IJG
+JPEG Library README, Windows Media SDK EULA, or QuickTime SDK EULA, the
+licensors of this Program grant you additional
+permission to convey the resulting work. Corresponding Source for a
+non-source form of such a combination shall include the source code for
+the parts of OpenSSL and IJG JPEG Library used as well as that of the covered
+work.
+
+You can contact Cyan Worlds, Inc. by email legal@cyan.com
+ or by snail mail at:
+      Cyan Worlds, Inc.
+      14617 N Newport Hwy
+      Mead, WA   99021
+
+*==LICENSE==*/
+
+#import "PLSPatcher.h"
+
+#include <string_theory/format>
+#include "StringTheory_NSString.h"
+
+#include "HeadSpin.h"
+#include "hsTimer.h"
+
+#include "pfPatcher/pfPatcher.h"
+#include "pfPatcher/plManifests.h"
+#include "plFileSystem.h"
+#include "plNetGameLib/plNetGameLib.h"
+
+class Patcher
+{
+   public:
+    PLSPatcher *parent;
+    void IOnPatchComplete(ENetError result, const ST::string &msg);
+    void IOnProgressTick(uint64_t curBytes, uint64_t totalBytes, const ST::string &status);
+    void IOnDownloadBegin(const plFileName &file);
+};
+
+@interface PLSPatcher ()
+@property BOOL selfPatched;
+@property pfPatcher *patcher;
+@property NSTimer *networkPumpTimer;
+@property Patcher cppPatcher;
+@end
+
+@implementation PLSPatcher
+
+- (id)init
+{
+    self = [super init];
+    self.selfPatched = false;
+
+    _cppPatcher.parent = self;
+
+    self.patcher = new pfPatcher();
+    _patcher->OnFileDownloadBegin(
+        std::bind(&Patcher::IOnDownloadBegin, _cppPatcher, std::placeholders::_1));
+    _patcher->OnProgressTick(std::bind(&Patcher::IOnProgressTick, _cppPatcher,
+                                       std::placeholders::_1, std::placeholders::_2,
+                                       std::placeholders::_3));
+    _patcher->OnCompletion(std::bind(&Patcher::IOnPatchComplete, _cppPatcher, std::placeholders::_1,
+                                     std::placeholders::_2));
+    _patcher->OnFileDownloadDesired(IApproveDownload);
+    _patcher->OnSelfPatch([&](const plFileName &file) {});
+    _patcher->OnRedistUpdate([&](const plFileName &file) {});
+
+    self.networkPumpTimer = [NSTimer timerWithTimeInterval:1.0 / 1000.0
+                                                   repeats:true
+                                                     block:^(NSTimer *_Nonnull timer) {
+                                                         hsTimer::IncSysSeconds();
+                                                         NetClientUpdate();
+                                                     }];
```

Comment:
I don't want to bikeshed formatting too much, but clang-format uses *very* wide visual indents in some places. Here for example it adds almost 50 spaces for a single indent level, which is excessive IMO and is quite inconvenient with smaller window/screen sizes.

Is there any way to make it wrap differently so that it needs less horizontal space, or perhaps use fixed-width indents instead of visual alignment over a certain width?

Reply:
It's aligning the `:`s which is convention for Obj-C. I'm not sure the is an alternative break style for Obj-C.

Reply:
WebKit's style seems to be to keep things on 1 line as much as possible (which I agree is terrible for readability). Ex: https://github.com/WebKit/WebKit/blob/875c04f6354f5201170ccd9fca83c4af0dff5ff1/Tools/MiniBrowser/mac/BrowserWindowController.m#L278-L281

Reply:
Yes. Million character long function calls on the same line is the other accepted Obj-C style. :D

Reply:
Wow, that's awful. I just checked, and clang-format really doesn't have *any* formatting options for Objective-C method calls... If these are our only choices, then I would prefer a single long line as well - at least that way my editor's soft wrapping can do its job properly and doesn't have to squeeze everything into a small column on the right.

I'm surprised that there's no way to make clang-format do something like this:

```suggestion
    self.networkPumpTimer = [NSTimer
        timerWithTimeInterval:1.0 / 1000.0
        repeats:true
        block:^(NSTimer *_Nonnull timer) {
            hsTimer::IncSysSeconds();
            NetClientUpdate();
        }];
```

Reply:
Would [`ObjCBlockIndentWidth`](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#objcblockindentwidth) perhaps work to force at least the blocks to not be indented so much? Or is that indent *on top* of any method argument indents? :/

Reply:
I'm hoping disabling of line length enforcement in Obj-C files should resolve this (although I have not made changes in the source files.)

That might cause more style discussions in PRs. But it should allow developers to use judgement on if long Obj-C lines should wrap or be broken up.

I experimented with just allowing longer lines in Obj-C files and was dissatisfied with the results. (Especially in since Clang Format will push code both back onto single lines or multiple lines.)

Reply:
Xcode's format rules produces the following output for that function:

```
    self.networkPumpTimer = [NSTimer timerWithTimeInterval:1.0 / 1000.0
                                                   repeats:true
                                                     block:^(NSTimer* _Nonnull timer) {
        hsTimer::IncSysSeconds();
        NetClientUpdate();
    }];
```

I think Xcode's indentation is more sensible - and I'm not looking forward to having Xcode's Obj-C formatter and Clang-Format fighting all the time. I'm going to see if I can set up Clang-Format's Obj-C rules to match Xcode's Obj-C rules for this case.

Reply:
Sigh... Apparently Clang-Format added this for C++ Lambdas but not Obj-C blocks. There is a bug here:
https://github.com/llvm/llvm-project/issues/53740

Until this is fixed one solution could be... Don't rely on Clang-Format for actual enforcement? I think what Xcode products is acceptable, and I would be ok if the code diverged from what Clang-Format does until Clang-Format supports this feature.

Reply:
> I think Xcode's indentation is more sensible

Yeah, that's definitely better for the block bodies. The large visual indents in method calls are the established Objective-C style, so that's fine (though personally I'm still not a fan, heh).